File: worker/src/sakuga.js
// Accès à l'API Sakugabooru depuis le Worker.
// Le site ne renvoie aucun en-tête CORS : le navigateur ne peut pas l'appeler
// directement, c'est tout l'intérêt de passer par ici. Les médias (vignettes,
// mp4), eux, sont chargés en direct par la page — <img> et <video> ne sont pas
// soumis au CORS.

const BASE = "https://www.sakugabooru.com";
const UA = "amvauto/0.1 (+https://github.com/AbilanBalakumaran/amvauto)";

const TAG_ARTIST = 1;
const TAG_COPYRIGHT = 3;
const VIDEO_EXTS = new Set(["mp4", "webm"]);

/* ---- Ce qui n'est pas du matériau de montage -----------------------------

   Un AMV de concours se juge aussi sur la propreté de ses sources : un carton-
   titre, une comparaison côte à côte, une capture d'écran de site sont des
   pollutions visuelles qui disqualifient un montage avant même qu'on regarde le
   rythme.

   AUCUN tag de sous-titre, de watermark ou d'incrustation TV n'existe sur
   Sakugabooru — vérifié le 22/09/2026 : « subtitled », « watermark », « credits »,
   « broadcast_screen », « lower_third », « hardsub », « tv_broadcast » rendent
   tous une liste vide. Le site n'indexe pas ça.

   Ce qu'il indexe, en revanche, et qui porte du texte ou n'est pas de l'animation
   jouable :

     comparison, genga_comparison   deux images côte à côte, avec des libellés
     title_animation                un carton-titre : du texte à l'écran
     screencap                      une capture d'écran, pas un cut
     live_action, stop_motion       ce n'est pas de l'animation dessinée
     tagme                          personne n'a su dire ce que c'est

   Et le matériel de production, déjà écarté en partie : la liste était de quatre
   tags, il en existe seize. Un « settei » ou un « color_script » n'est pas plus
   montable qu'un « genga ». */
export const PAPER_TAGS = new Set([
  // Le papier de production.
  "genga", "production_materials", "layout", "douga", "settei", "storyboard",
  "character_design", "background_design", "concept_art", "color_script",
  "timesheet", "rough", "shiage", "illustration", "sprite", "cel", "flipbook",
  // Ce qui porte du texte, ou n'est pas de l'animation.
  "comparison", "genga_comparison", "title_animation", "screencap",
  "live_action", "stop_motion", "tagme",
]);

/* ---- Le cadre : une ligne de temps 16:9 sans barres noires ----------------

   Le rendu est en 16:9. Un cut en 4:3 y entre avec deux bandes noires sur les
   côtés — ce n'est pas une déformation, le rendu ne déforme jamais, mais ça se
   voit et ça casse la ligne.

   RELEVÉ SUR LE VRAI CATALOGUE, et c'est ce qui décide de la règle : sur trois
   cents cuts vidéo de Naruto Shippuden, Chainsaw Man et Jujutsu Kaisen, les
   ratios sont 1,78 et 1,77 à 95 % — du 16:9 — avec quelques 1,85, un 1,33 et
   trois 2,40. On garde donc la fenêtre du 16:9 élargie au cinéma, et l'on écarte
   le 4:3 et le cinémascope.

   La définition, elle, NE PEUT PAS être filtrée au-dessus de 720p : sur ces trois
   cents cuts, CENT POUR CENT sont entre 480 et 719 lignes, et aucun n'atteint
   720. Sakugabooru sert des extraits volontairement légers. Exiger 720p viderait
   le catalogue entier. Le plancher est donc posé là où il protège sans rien
   casser : sous 400 lignes, ce n'est plus une source montable. */
const RATIO_MINI = 1.6;
const RATIO_MAXI = 1.9;
const LIGNES_MINI = 400;

export function cadrePropre(post) {
  const h = post.height || 0;
  const w = post.width || 0;
  if (h < LIGNES_MINI) return false;
  if (!h || !w) return true;   // sans définition connue, on ne présume rien
  const ratio = w / h;
  return ratio >= RATIO_MINI && ratio <= RATIO_MAXI;
}

// Table des animateurs : lourde à récupérer, stable dans le temps. On la garde
// dans l'isolate et on laisse le cache Cloudflare absorber le reste.
let artistCache = null;

/* ---- Ce que Sakugabooru NE dit pas : les personnages ----------------------

   Un AMV a besoin d'un fil : sur mille huit cents rushs, enchaîner des plans qui
   n'ont personne en commun donne une bande-démo d'animateurs, pas un récit. Le
   fil évident serait le personnage — suivre Naruto, ou monter un duel Naruto
   contre Sasuke.

   Sakugabooru ne le permet pas. Le site n'étiquette AUCUN personnage :
   « naruto_uzumaki », « sasuke_uchiha », « gojo_satoru » ne sont pas des tags,
   et le type 4 de son catalogue — celui que Danbooru réserve aux personnages —
   ne compte que dix tags, qui sont des signatures d'animation : kanada_light_flare,
   itano_circus, obari_punch, ebata_walk. Relevé sur le site le 21/09/2026 :
   tag.json?name=naruto_uzumaki rend une liste vide.

   La seule identité que porte un cut, c'est son ANIMATEUR — et celle-là est
   riche : « hiroyuki_yamashita », « shingo_yamashita », plusieurs par plan,
   quinze cents pour Hironori Tanaka. C'est elle qui sert de fil. Ce n'est pas la
   continuité de personnage demandée, c'est la continuité de main : le geste, le
   trait, la façon de déformer. En AMV de compétition, c'est un fil reconnu — le
   sakuga showcase se monte comme ça. */

async function api(path, params, ttl) {
  const url = new URL(BASE + path);
  for (const [key, value] of Object.entries(params)) url.searchParams.set(key, value);
  const response = await fetch(url, {
    headers: { "User-Agent": UA, Accept: "application/json" },
    cf: { cacheTtl: ttl, cacheEverything: true },
  });
  if (!response.ok) throw new Error(`Sakugabooru a répondu ${response.status}`);
  return response.json();
}

export async function artistTags() {
  if (artistCache) return artistCache;
  const tags = await api("/tag.json", { type: TAG_ARTIST, order: "count", limit: 2000 }, 86400);
  artistCache = new Set(tags.map((tag) => tag.name).filter((name) => name !== "artist_unknown"));
  return artistCache;
}

export async function searchSeries(query) {
  const pattern = query.trim().toLowerCase().replace(/\s+/g, "_");
  if (!pattern) return [];
  let tags = await api("/tag.json", { name: pattern, type: TAG_COPYRIGHT, order: "count", limit: 30 }, 3600);
  if (!tags.length && pattern.includes("_")) {
    tags = await api("/tag.json", { name: pattern.split("_")[0], type: TAG_COPYRIGHT, order: "count", limit: 30 }, 3600);
  }
  return tags.filter((tag) => tag.count > 0).sort((a, b) => b.count - a.count);
}

const PAGE_SIZE = 100;

const VAGUE = 5;

export async function posts(tags, limit) {
  // L'API plafonne à 100 posts par page. Au-delà on pagine, par vagues de cinq :
  // en lancer vingt d'un coup sur un site communautaire serait grossier, et une
  // page vide signifie qu'on a atteint le fond du tag.
  const pages = Math.max(1, Math.ceil(limit / PAGE_SIZE));
  const raw = [];

  /* La première page part seule. Un tag sans vidéo — le cas de figure quand on
     essaie plusieurs candidats avant de trouver le bon — coûtait cinq requêtes
     pour apprendre qu'il n'y a rien : la vague était lancée d'un bloc. Or le
     plan gratuit plafonne à cinquante sous-requêtes par appel, et six tags
     essayés en dépassaient. Une requête suffit à le savoir. */
  const premiere = await api("/post.json", { tags, limit: PAGE_SIZE, page: 1 }, 1800);
  raw.push(...premiere);
  if (premiere.length === PAGE_SIZE) {
    for (let debut = 1; debut < pages; debut += VAGUE) {
      const vague = Array.from({ length: Math.min(VAGUE, pages - debut) }, (_, index) =>
        api("/post.json", { tags, limit: PAGE_SIZE, page: debut + index + 1 }, 1800),
      );
      const lots = await Promise.all(vague);
      raw.push(...lots.flat());
      if (lots.some((lot) => lot.length < PAGE_SIZE)) break;
    }
  }
  const artists = await artistTags();
  return raw.map((item) => {
    const tagList = (item.tags || "").split(" ").filter(Boolean);
    return {
      id: item.id,
      tags: tagList,
      score: item.score || 0,
      file_url: item.file_url || "",
      file_ext: item.file_ext || "",
      file_size: item.file_size || 0,
      preview_url: item.preview_url || "",
      sample_url: item.sample_url || "",
      width: item.width || 0,
      height: item.height || 0,
      rating: item.rating || "s",
      source: item.source || "",
      /* Les tags bruts, pas les noms rendus lisibles : c'est sur eux que le
         montage compare, et « Hiroyuki Yamashita » ne se recompare plus à
         « hiroyuki_yamashita ». « artist_unknown » est déjà écarté par la table :
         ce n'est pas une main, c'est une absence de crédit. */
      artists: tagList.filter((tag) => artists.has(tag)),
    };
  });
}

// Ce qui est réellement montable : une vidéo, pas du papier, pas de contenu
// explicite.
export const montables = (found) =>
  found.filter(
    (post) =>
      VIDEO_EXTS.has(post.file_ext) &&
      !post.tags.some((tag) => PAPER_TAGS.has(tag)) &&
      cadrePropre(post) &&
      post.rating !== "e",
  );

// Les cuts d'une série.
export async function rushes(seriesTag, limit = 1000) {
  return montables(await posts(`${seriesTag} order:score`, limit));
}

/* Les meilleurs cuts du site, sans série imposée.

   C'est ce que demande un AMV mixte : ne pas choisir les animés soi-même, et
   laisser remonter ce que la communauté a le mieux noté. Un tag d'ambiance
   restreint la pioche sans la fermer — « fighting order:score » rend les
   meilleurs plans de combat, d'où qu'ils viennent. */
export async function rushesPartout(tagAmbiance, limit = 1000) {
  const requete = tagAmbiance ? `${tagAmbiance} order:score` : "order:score";
  return montables(await posts(requete, limit));
}

/* La série d'un post, lue dans ses tags.

   En mixte, l'animé n'est plus une donnée d'entrée : c'est une donnée de
   sortie. Sans elle, le montage afficherait deux cents plans sans savoir d'où
   ils viennent, et le chapitrage par animé n'aurait plus de sens. */
let copyrightCache = null;
export async function copyrightTags() {
  if (copyrightCache) return copyrightCache;
  const tags = await api("/tag.json", { type: TAG_COPYRIGHT, order: "count", limit: 2000 }, 86400);
  copyrightCache = new Set(tags.map((tag) => tag.name));
  return copyrightCache;
}

export function serieDe(post, series) {
  const tag = post.tags.find((nom) => series.has(nom));
  return tag ? tag.replace(/_/g, " ").replace(/\b\w/g, (c) => c.toUpperCase()) : "";
}

File: amvauto/sakuga.py
"""Client HTTP minimal pour l'API Sakugabooru (moebooru).

Sakugabooru indexe des « cuts » d'animation : de courts extraits vidéo isolés
d'un episode, taggés par serie, par animateur et par type d'animation. C'est la
source de rushs la plus directement exploitable pour monter un AMV.
"""

from __future__ import annotations

import json
import time
import urllib.parse
import urllib.request
from dataclasses import dataclass, field
from typing import Iterable

BASE_URL = "https://www.sakugabooru.com"
USER_AGENT = "amvauto/0.1 (+https://github.com/AbilanBalakumaran/amvauto)"

# Types de tags moebooru.
TAG_GENERAL = 0
TAG_ARTIST = 1
TAG_COPYRIGHT = 3
TAG_CHARACTER = 4

VIDEO_EXTS = {"mp4", "webm"}


class SakugaError(RuntimeError):
    pass


@dataclass
class Post:
    """Un cut d'animation."""

    id: int
    tags: list[str]
    score: int
    file_url: str
    file_ext: str
    file_size: int
    preview_url: str
    width: int
    height: int
    rating: str
    source: str
    created_at: int
    artists: list[str] = field(default_factory=list)
    series: list[str] = field(default_factory=list)

    @property
    def page_url(self) -> str:
        return f"{BASE_URL}/post/show/{self.id}"

    @property
    def is_video(self) -> bool:
        return self.file_ext in VIDEO_EXTS

    @property
    def is_production_material(self) -> bool:
        """Genga / layouts : du papier, pas du rush montable."""
        return bool({"genga", "production_materials", "layout", "douga"} & set(self.tags))

    @classmethod
    def from_json(cls, raw: dict) -> "Post":
        return cls(
            id=raw["id"],
            tags=raw.get("tags", "").split(),
            score=raw.get("score", 0),
            file_url=raw.get("file_url", ""),
            file_ext=raw.get("file_ext", ""),
            file_size=raw.get("file_size", 0),
            preview_url=raw.get("preview_url", ""),
            width=raw.get("width", 0),
            height=raw.get("height", 0),
            rating=raw.get("rating", "s"),
            source=raw.get("source", "") or "",
            created_at=raw.get("created_at", 0),
        )


class SakugaClient:
    def __init__(self, base_url: str = BASE_URL, delay: float = 0.4, timeout: int = 30):
        self.base_url = base_url.rstrip("/")
        self.delay = delay
        self.timeout = timeout
        self._last_call = 0.0
        self._tag_types: dict[str, int] = {}

    # -- transport ---------------------------------------------------------

    def _get(self, path: str, params: dict) -> list | dict:
        wait = self.delay - (time.monotonic() - self._last_call)
        if wait > 0:
            time.sleep(wait)
        url = f"{self.base_url}{path}?{urllib.parse.urlencode(params)}"
        req = urllib.request.Request(url, headers={"User-Agent": USER_AGENT})
        try:
            with urllib.request.urlopen(req, timeout=self.timeout) as resp:
                payload = json.loads(resp.read().decode("utf-8"))
        except Exception as exc:  # réseau, JSON, HTTP…
            raise SakugaError(f"appel {path} en échec : {exc}") from exc
        finally:
            self._last_call = time.monotonic()
        return payload

    # -- tags --------------------------------------------------------------

    def search_tags(self, pattern: str, tag_type: int | None = None, limit: int = 40) -> list[dict]:
        """Cherche des tags par motif (utilisé pour résoudre un nom d'animé)."""
        params = {"name": pattern, "limit": limit, "order": "count"}
        if tag_type is not None:
            params["type"] = tag_type
        tags = self._get("/tag.json", params)
        return [t for t in tags if t.get("count", 0) > 0]

    def top_tags(self, tag_type: int, limit: int = 500) -> list[dict]:
        return self._get("/tag.json", {"type": tag_type, "order": "count", "limit": limit})

    def load_tag_types(self, types: Iterable[int] = (TAG_ARTIST,), limit: int = 2000) -> None:
        """Pré-charge une table nom -> type, pour distinguer animateur et série."""
        for tag_type in types:
            for tag in self.top_tags(tag_type, limit=limit):
                self._tag_types[tag["name"]] = tag_type

    def tag_type(self, name: str) -> int:
        return self._tag_types.get(name, TAG_GENERAL)

    def resolve_series(self, query: str) -> list[dict]:
        """Nom d'animé libre -> tags de série candidats, le plus fourni d'abord."""
        pattern = query.strip().lower().replace(" ", "_")
        candidates = self.search_tags(pattern, tag_type=TAG_COPYRIGHT)
        if not candidates and "_" in pattern:
            # « chainsaw man reze » ne matche rien : on retente sur le 1er mot.
            candidates = self.search_tags(pattern.split("_")[0], tag_type=TAG_COPYRIGHT)
        return sorted(candidates, key=lambda t: -t.get("count", 0))

    # -- posts -------------------------------------------------------------

    def posts(self, tags: str, limit: int = 40, page: int = 1) -> list[Post]:
        raw = self._get("/post.json", {"tags": tags, "limit": limit, "page": page})
        posts = [Post.from_json(item) for item in raw]
        for post in posts:
            post.artists = [
                t for t in post.tags
                if self.tag_type(t) == TAG_ARTIST and t != "artist_unknown"
            ]
            post.series = [t for t in post.tags if self.tag_type(t) == TAG_COPYRIGHT]
        return posts

    def rushes(self, series_tag: str, limit: int = 40, extra_tags: str = "") -> list[Post]:
        """Les cuts les mieux notés d'une série, filtrés sur ce qui est montable."""
        query = f"{series_tag} order:score"
        if extra_tags:
            query = f"{query} {extra_tags}"
        found = self.posts(query, limit=limit)
        return [p for p in found if p.is_video and not p.is_production_material]
